From 85e6bd72d87fafaeab39461f5a5ddaa54f28216a Mon Sep 17 00:00:00 2001
From: Ben Gainey <ben.gainey@arm.com>
Date: Thu, 9 Nov 2023 11:41:27 +0000
Subject: [RFC PATCH 2/3] arm_pmuv3: Add config bits for sample period strobing

To expose the ability to alternate between sample periods to tools.
The field `strobe_period` is defined for config2 to hold the alternate
sample period. A non-zero value will enable strobing.

Signed-off-by: Ben Gainey <ben.gainey@arm.com>
---
 arch/arm64/kernel/perf_event.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/arch/arm64/kernel/perf_event.c b/arch/arm64/kernel/perf_event.c
index b4044469527e0..eceb130e7a506 100644
--- a/arch/arm64/kernel/perf_event.c
+++ b/arch/arm64/kernel/perf_event.c
@@ -285,15 +285,27 @@ static const struct attribute_group armv8_pmuv3_events_attr_group = {
 
 PMU_FORMAT_ATTR(event, "config:0-15");
 PMU_FORMAT_ATTR(long, "config1:0");
+PMU_FORMAT_ATTR(strobe_period, "config2:0-31");
 
 static inline bool armv8pmu_event_is_64bit(struct perf_event *event)
 {
 	return event->attr.config1 & 0x1;
 }
 
+static inline u32 armv8pmu_event_strobe_period(struct perf_event *event)
+{
+	return event->attr.config2;
+}
+
+static inline bool armv8pmu_event_want_strobe(struct perf_event *event)
+{
+	return armv8pmu_event_strobe_period(event) != 0;
+}
+
 static struct attribute *armv8_pmuv3_format_attrs[] = {
 	&format_attr_event.attr,
 	&format_attr_long.attr,
+	&format_attr_strobe_period.attr,
 	NULL,
 };
 
@@ -996,6 +1008,16 @@ static int __armv8_pmuv3_map_event(struct perf_event *event,
 	if (armv8pmu_event_is_64bit(event))
 		event->hw.flags |= ARMPMU_EVT_64BIT;
 
+	/*
+	 * Support alternating between two sample periods
+	 */
+	if (armv8pmu_event_want_strobe(event)) {
+		u32 strobe_period = armv8pmu_event_strobe_period(event);
+		armpmu_set_strobe_period(&(event->hw), strobe_period);
+	} else {
+		armpmu_set_strobe_period(&(event->hw), 0);
+	}
+
 	/* Only expose micro/arch events supported by this PMU */
 	if ((hw_event_id > 0) && (hw_event_id < ARMV8_PMUV3_MAX_COMMON_EVENTS)
 	    && test_bit(hw_event_id, armpmu->pmceid_bitmap)) {
-- 
2.43.0

